app/src/main/java/gateway/controller/CtrlShareListWithWho.java

```diff
@@ -0,0 +1,83 @@
+package gateway.controller;
+
+import gateway.config.Config;
+import gateway.services.ServiceAuth;
+import gateway.services.ServiceAuth.ResUsername;
+import gateway.services.UtilValidator;
+import gateway.soap.request.ReqFile;
+import gateway.soap.response.ResShareListWithWho;
+import gateway.soap.response.ResStatus;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.UUID;
+import org.json.JSONArray;
+import org.json.JSONObject;
+
+public class CtrlShareListWithWho
+{
+	public static ResShareListWithWho share_list_with_who (ReqFile args)
+	{
+		ResShareListWithWho resShareListWithWho = new ResShareListWithWho ();
+
+		// Validate Field
+		ResStatus resValidate = UtilValidator.validate (args);
+		if (resValidate.error) {
+			return ResStatus.downCast (ResShareListWithWho.class, resValidate);
+		}
+
+		// User authentication
+		ResStatus resAuth = ServiceAuth.authenticate (args.token);
+		if (resAuth.error) {
+			return ResStatus.downCast (ResShareListWithWho.class, resAuth);
+		}
+
+		String url = Config.getMetadataBaseUrl () + "/files/shared_with_who/" + args.fileUUID;
+
+		try {
+			// Make an HTTP GET request
+			HttpResponse<String> response = HttpClient.newHttpClient ().send (
+				HttpRequest.newBuilder ().uri (URI.create (url)).GET ().build (),
+				HttpResponse.BodyHandlers.ofString ());
+
+			// Parse the JSON response
+			JSONObject responseBody = new JSONObject (response.body ());
+			resShareListWithWho.code = response.statusCode ();
+
+			if (resShareListWithWho.code == 200) {
+				// If the response code is 200, process the received files
+				JSONArray sharedWithArray = responseBody.getJSONArray ("shared_with");
+				String[] usernames = new String[sharedWithArray.length ()];
+
+				for (int i = 0; i < sharedWithArray.length (); i++) {
+					// Process each shared_with in the response JSON
+					String sharedWithObject = sharedWithArray.getString (i);
+
+					// Get username from UUID
+					ResUsername username =
```

Comment:
Missing error check from `ResUsername`. If a username returns error don't add it to the array. Make sure to not leave holes in the array.

As a suggestion you can use a List and use a method similar to `append` to add the items. Then at the end you can convert it to Array using a method called `to_array` I think. 

This way you don't have to set the usernames to the array using indexes.